mailpoet/lib/Automation/Integrations/WooCommerce/Fields/TermOptionsBuilder.php

```diff
@@ -18,28 +21,58 @@ public function __construct(
 
   /** @return array<array{id: int, name: string}> */
   public function getTermOptions(string $taxonomy): array {
+    if (isset($this->terms[$taxonomy])) {
+      return $this->terms[$taxonomy];
+    }
+
+    $taxonomyObject = $this->wordPress->getTaxonomy($taxonomy);
+    if (!$taxonomyObject) {
+      $this->terms[$taxonomy] = [];
+      return $this->terms[$taxonomy];
+    }
+
     $terms = $this->wordPress->getTerms(['taxonomy' => $taxonomy, 'hide_empty' => false, 'orderby' => 'name']);
     if ($terms instanceof WP_Error) {
-      return [];
+      $this->terms[$taxonomy] = [];
+      return $this->terms[$taxonomy];
     }
-    return $this->buildTermsList((array)$terms);
+    $this->terms[$taxonomy] = $this->buildTermsList((array)$terms, $taxonomyObject);
+    return $this->terms[$taxonomy];
   }
 
   /** @return array<array{id: int, name: string}> */
-  private function buildTermsList(array $terms, int $parentId = 0): array {
-    $parents = array_filter($terms, function ($term) use ($parentId) {
-      return $term instanceof WP_Term && $term->parent === $parentId;
-    });
-
-    usort($parents, function (WP_Term $a, WP_Term $b) {
-      return $a->name <=> $b->name;
-    });
+  private function buildTermsList(array $terms, \WP_Taxonomy $taxonomy, int $parentId = 0, array $lookup = []): array {
+    if (!$taxonomy->hierarchical) {
+      return array_map(
+        function ($term) {
+          return ['id' => (int)$term->term_id, 'name' => $term->name]; //phpcs:ignore: Squiz.NamingConventions.ValidVariableName.MemberNotCamelCaps
+        },
+        $terms
+      );
+    }
+    if ($lookup === []) {
+      foreach ($terms as $term) {
+        if (!$term instanceof WP_Term) {
+          continue;
+        }
+        $lookup[$term->parent][] = $term;
+      }
+    }
+    if (empty($lookup[$parentId])) {
+      return [];
+    }
 
     $list = [];
-    foreach ($parents as $term) {
+    foreach ($lookup[$parentId] as $term) {
+      if (!$term instanceof WP_Term || $term->parent !== $parentId) {
+        continue;
+      }
       $id = $term->term_id; // phpcs:ignore Squiz.NamingConventions.ValidVariableName.MemberNotCamelCaps
       $list[] = ['id' => (int)$id, 'name' => $term->name];
-      foreach ($this->buildTermsList($terms, $id) as $child) {
+      if (empty($lookup[$id])) {
```

Comment:
I think `empty()` is not pretty slow and wasn't recommended to be used within A8C, `isset` is probably better.